Objects/interpolationobject.c

```diff
@@ -0,0 +1,217 @@
+    PyObject *conversion;
+    PyObject *format_spec;
+} interpolationobject;
+
+/*[clinic input]
+@classmethod
+Interpolation.__new__ as interpolation_new
+
+    value: object
+    expression: object(subclass_of='&PyUnicode_Type')
+    conversion: object(converter='_conversion_converter') = None
+    format_spec: object(subclass_of='&PyUnicode_Type', c_default='&_Py_STR(empty)') = ""
+[clinic start generated code]*/
+
+static PyObject *
+interpolation_new_impl(PyTypeObject *type, PyObject *value,
+                       PyObject *expression, PyObject *conversion,
+                       PyObject *format_spec)
+/*[clinic end generated code: output=6488e288765bc1a9 input=d91711024068528c]*/
+{
+    interpolationobject *self = (interpolationobject *) type->tp_alloc(type, 0);
+    if (!self) {
+        return NULL;
+    }
+
+    self->value = Py_NewRef(value);
+    self->expression = Py_NewRef(expression);
+    self->conversion = Py_NewRef(conversion);
+    self->format_spec = Py_NewRef(format_spec);
+    return (PyObject *) self;
+}
+
+static void
+interpolation_dealloc(interpolationobject *self)
+{
+    PyObject_GC_UnTrack(self);
+    Py_CLEAR(self->value);
+    Py_CLEAR(self->expression);
+    Py_CLEAR(self->conversion);
+    Py_CLEAR(self->format_spec);
+    Py_TYPE(self)->tp_free(self);
+}
+
+static int
+interpolation_traverse(interpolationobject *self, visitproc visit, void *arg)
+{
+    Py_VISIT(self->value);
+    Py_VISIT(self->expression);
+    Py_VISIT(self->conversion);
+    Py_VISIT(self->format_spec);
+    return 0;
+}
+
+static PyObject *
+interpolation_repr(interpolationobject *self)
+{
+    return PyUnicode_FromFormat("%s(%R, %R, %R, %R)",
+                                _PyType_Name(Py_TYPE(self)),
+                                self->value, self->expression,
+                                self->conversion, self->format_spec);
+}
+
+static PyMemberDef interpolation_members[] = {
+    {"value", Py_T_OBJECT_EX, offsetof(interpolationobject, value), Py_READONLY, "Value"},
+    {"expression", Py_T_OBJECT_EX, offsetof(interpolationobject, expression), Py_READONLY, "Expression"},
+    {"conversion", Py_T_OBJECT_EX, offsetof(interpolationobject, conversion), Py_READONLY, "Conversion"},
+    {"format_spec", Py_T_OBJECT_EX, offsetof(interpolationobject, format_spec), Py_READONLY, "Format specifier"},
+    {NULL}
+};
+
+PyTypeObject _PyInterpolation_Type = {
+    PyVarObject_HEAD_INIT(NULL, 0)
+    .tp_name = "string.templatelib.Interpolation",
+    .tp_doc = PyDoc_STR("Interpolation object"),
+    .tp_basicsize = sizeof(interpolationobject),
+    .tp_itemsize = 0,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    .tp_new = (newfunc) interpolation_new,
+    .tp_alloc = PyType_GenericAlloc,
+    .tp_dealloc = (destructor) interpolation_dealloc,
+    .tp_free = PyObject_GC_Del,
+    .tp_repr = (reprfunc) interpolation_repr,
+    .tp_members = interpolation_members,
+    .tp_traverse = (traverseproc) interpolation_traverse,
```

Comment:
```suggestion
    .tp_new = interpolation_new,
    .tp_alloc = PyType_GenericAlloc,
    .tp_dealloc = interpolation_dealloc,
    .tp_free = PyObject_GC_Del,
    .tp_repr = interpolation_repr,
    .tp_members = interpolation_members,
    .tp_traverse = interpolation_traverse,
```

Let's avoid introducing more UBs from now on :)